ardour: add a videoSupport option, harvid: init at 0.8.3, xjadeo: init at 0.8.10

mitchmindtree opened the pull request:
##### Motivation

### Video support in Adour with `videoSupport`

This adds a `bool` option that, when set to `true`, enables video
timeline support for the Ardour DAW. This is commonly useful for
soundtrack composition, sound design for film, etc.

When enabled, `videoSupport` ensures that both `harvid` and `xjadeo` are
available to the `ardour6` exe via the PATH. `harvid` decodes the video
stream in real-time to produce still images (I think for thumbnail
support for the timeline?). `xjadeo` acts as a video monitoring window
that whose playback position is synchronised to the Ardour playhead.

`videoSupport` remains disabled by default, preserving the original
behaviour.

Video support can be added to ardour in your system or home
configuration package list with:

```
(ardour.override { videoSupport = true; })
```

### `harvid` and `xjadeo`

`harvid` decodes still images from movie files and serves them via HTTP.
`xjadeo` is an "X Jack Video Monitor".
Both executables must be available via `PATH` in order for ardour's video
timeline support to work. You can read more about both tools in their
respective `meta` attributes.

Tested builds of ardour both with and without `videoSupport` and both
appear to work nicely. Both `harvid` and `xjadeo` also appear to work
as intended from their respective CLIs.

---

##### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

cc @cillianderoiste  @magnetophon, current maintainers of the ardour package.

mitchmindtree committed:
maintainers: add mitchmindtree

In preparation for the following commits adding ardour-related video
tooling and new `videoSupport` option.

mitchmindtree committed:
harvid: init at 0.8.3

harvid decodes still images from movie files and serves them via HTTP.
Please refer to the meta attribute for more information.

This commit is part of a PR aimed at enabling video support in ardour.

mitchmindtree committed:
xjadeo: init at 0.8.10

xjadeo is the X Jack Video Monitor.

Please refer to the meta attribute for more information on xjadeo.

This commit is part of a PR aimed at enabling video support in ardour.

mitchmindtree committed:
ardour: add a `videoSupport` option

This adds a `bool` option that, when set to `true`, enables video
timeline support for the Ardour DAW. This is commonly useful for
soundtrack composition, sound design for film, etc.

When enabled, `videoSupport` ensures that both `harvid` and `xjadeo` are
available to the `ardour6` exe via the PATH. `harvid` decodes the video
stream in real-time to produce still images (I think for thumbnail
support for the timeline?). `xjadeo` acts as a video monitoring window
that whose playback position is synchronised to the Ardour playhead.

`videoSupport` remains disabled by default, preserving the original
behaviour.

Video support can be added to ardour in your system or home
configuration package list with:

```
(ardour.override { videoSupport = true; })
```

r-rmcgibbo commented:
Result of `nixpkgs-review pr 115121` at 65ee58f1 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li><a href="https://gist.github.com/efb57778702300352d6b55b43bce3863">ardour</a></li></li>
    <li><a href="https://gist.github.com/2e0abb1a731336abc51fb00b5a4c2f33">harvid</a></li></li>
    <li><a href="https://gist.github.com/1d4b9d61ccc3663508d4cdd6c376a512">xjadeo</a></li></li>
  </ul>
</details>
<details>
  <summary>5 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/maintainers-missing.md">maintainers-missing</a>

Package does not have a maintainer. Consider adding yourself?

Near pkgs/tools/video/xjadeo/default.nix:33:3:
```
   |
33 |   meta = with lib; {
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/tools/video/harvid/default.nix:17:3:
```
   |
17 |   installPhase = ''
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`gpl2` is a deprecated license, check if project uses `gpl2Plus` or `gpl2Only` and change `meta.license` accordingly.

Near pkgs/applications/audio/ardour/default.nix:189:5:
```
    |
189 |     license = licenses.gpl2;
    |     ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`gpl2` is a deprecated license, check if project uses `gpl2Plus` or `gpl2Only` and change `meta.license` accordingly.

Near pkgs/tools/video/harvid/default.nix:36:5:
```
   |
36 |     license = licenses.gpl2;
   |     ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`gpl2` is a deprecated license, check if project uses `gpl2Plus` or `gpl2Only` and change `meta.license` accordingly.

Near pkgs/tools/video/xjadeo/default.nix:42:5:
```
   |
42 |     license = licenses.gpl2;
   |     ^
```
</li>
  </ul>
</details>

SuperSandro2000 commented on pkgs/applications/audio/ardour/default.nix:
```diff
@@ -147,18 +152,25 @@ stdenv.mkDerivation rec {
+          "$out/share/icons/hicolor/''${size}x''${size}/apps/ardour6.png"
+      done
+      install -vDm 644 "ardour.1"* -t "$out/share/man/man1"
+    '';
```
```suggestion
    '' + lib.optionalString ''
      videoSupportScript = lib.optionalString videoSupport ''
      # `harvid` and `xjadeo` must be accessible in `PATH` for video to work.
      wrapProgram "$out/bin/ardour6" \
        --prefix PATH : "${lib.makeBinPath [ harvid xjadeo ]}"
    '';
```
This is far simpler than concatStrings.

mitchmindtree replied:
Thanks! I'd forgotten you could concatenate strings with `+`.

SuperSandro2000 commented on pkgs/applications/audio/ardour/default.nix:
```diff
@@ -147,18 +152,25 @@ stdenv.mkDerivation rec {
+  postInstall = let
+    script = ''
+      # wscript does not install these for some reason
+      install -vDm 644 "build/gtk2_ardour/ardour.xml" \
+        -t "$out/share/mime/packages"
+      install -vDm 644 "build/gtk2_ardour/ardour6.desktop" \
+        -t "$out/share/applications"
```
```suggestion
      install -vDm 644 "build/gtk2_ardour/ardour.xml" -t "$out/share/mime/packages"
      install -vDm 644 "build/gtk2_ardour/ardour6.desktop" -t "$out/share/applications"
```

mitchmindtree replied:
Going to leave these unrelated formatting changers for the original authors/maintainers who wrote this.

SuperSandro2000 commented on pkgs/tools/video/harvid/default.nix:
```diff
@@ -0,0 +1,42 @@
+{ lib, stdenv, fetchFromGitHub, ffmpeg, libjpeg, libpng, pkg-config }:
+stdenv.mkDerivation {
```
```suggestion
{ lib, stdenv, fetchFromGitHub, ffmpeg, libjpeg, libpng, pkg-config }:

stdenv.mkDerivation rec {
```

SuperSandro2000 commented on pkgs/tools/video/harvid/default.nix:
```diff
@@ -0,0 +1,42 @@
+  src = fetchFromGitHub {
+    owner = "x42";
+    repo = "harvid";
+    rev = "d71921ba52efd1362973df850213eebe97a7ecc6";
```
```suggestion
    rev = "v${version}";
```

SuperSandro2000 commented on pkgs/tools/video/harvid/default.nix:
```diff
@@ -0,0 +1,42 @@
+
+  installPhase = ''
+    runHook preInstall
+    make DESTDIR="$out" libdir="/lib" install
```
Please convert this into makeFlags and postInstall.

mitchmindtree replied:
Thanks for the tip!

SuperSandro2000 commented on pkgs/tools/video/harvid/default.nix:
```diff
@@ -0,0 +1,42 @@
+  '';
+
+  meta = with lib; {
+    description =
+      "Decodes still images from movie files and serves them via HTTP";
```
```suggestion
    description = "Decodes still images from movie files and serves them via HTTP";
```

mitchmindtree replied:
Omitting this formatting change as the current result is produced by `nixfmt`.

SuperSandro2000 replied:
nixfmt is not the official formatter for nixpkgs and we do not have a ~80 line length limit. It is totally fine to not linebreak a 120 or even 240 character URL and in this example the line break only stretches the file instead of making it more reable.

SuperSandro2000 commented on pkgs/tools/video/harvid/default.nix:
```diff
@@ -0,0 +1,42 @@
+      processor.
+    '';
+    homepage = "http://x42.github.io/harvid";
+    license = licenses.gpl2;
```
gpl2Only or gpl2Plus?

SuperSandro2000 commented on pkgs/tools/video/xjadeo/default.nix:
```diff
@@ -0,0 +1,47 @@
+{ lib, stdenv, autoreconfHook, fetchFromGitHub, ffmpeg, freetype, libGLU
+, libjack2, liblo, libX11, libXv, pkg-config, portmidi, xorg }:
+stdenv.mkDerivation {
```
```suggestion
, libjack2, liblo, libX11, libXv, pkg-config, portmidi, xorg }:

stdenv.mkDerivation rec {
```

SuperSandro2000 commented on pkgs/tools/video/xjadeo/default.nix:
```diff
@@ -0,0 +1,47 @@
+  src = fetchFromGitHub {
+    owner = "x42";
+    repo = "xjadeo";
+    rev = "c56548cad7b9bd358eb2c776de5ec5323f41006e";
```
```suggestion
    rev = "v${version}";
```

SuperSandro2000 commented on pkgs/tools/video/xjadeo/default.nix:
```diff
@@ -0,0 +1,47 @@
+    sha256 = "0cwh3zimrysyank3x1bcysdz59f8srqdca7jwv97ri3y63dlhifj";
+  };
+
+  nativeBuildInputs = [
```
Are those packages all required to build the package or are they runtime deps?

mitchmindtree replied:
They are listed under "build dependencies" in the README, but as mentioned in the comment the last few are recommended optional deps. Seeing as they're recommended I've included them for now.

On this topic, I'm not sure if these would be better split between `nativeBuildInputs` and `buildInputs`? The distinction between the two is still a bit fuzzy in my mind - I'll check the manual again.

SuperSandro2000 replied:
Everything that is executed the host goes into nativeBuildInputs.

SuperSandro2000 commented on pkgs/tools/video/xjadeo/default.nix:
```diff
@@ -0,0 +1,47 @@
+      synchronizing movie frames with external events.
+    '';
+    homepage = "http://xjadeo.sf.net";
+    license = licenses.gpl2;
```
gpl2Only or gpl2Plus?

mitchmindtree replied:
I'm unsure which is which to be honest!

Here's the xjadeo license:

https://github.com/x42/xjadeo/blob/master/COPYING

and here's the harvid one:

https://github.com/x42/harvid/blob/master/COPYING

any advice on which to use and I'll make the change.

Ardour currently has `gpl2`, and seeing as it is by the same group of authors and those COPYING pages were also GPL2 I thought I'd copy suit, but I understand it's better to be more specific when possible.

SuperSandro2000 replied:
It is plus https://github.com/x42/xjadeo/blob/master/src/xjadeo/main.c#L8

SuperSandro2000 commented on pkgs/tools/video/xjadeo/default.nix:
```diff
@@ -0,0 +1,47 @@
+      soundtrack composition, video monitoring or any task that requires to
+      synchronizing movie frames with external events.
+    '';
+    homepage = "http://xjadeo.sf.net";
```
```suggestion
    homepage = "http://xjadeo.sourceforge.net/";
```

SuperSandro2000 commented on pkgs/tools/video/xjadeo/default.nix:
```diff
@@ -0,0 +1,47 @@
+    maintainers = with maintainers; [ mitchmindtree ];
+  };
+}
+
```
One final new line to much.

mitchmindtree commented:
Thanks for the thorough review @SuperSandro2000!

I believe I've addressed what I can - the changes have been addressed via a rebase of the original commits. Let me know if these sorts of small changes are better addressed with follow-up commits - I assumed I was better off editing the existing commits to make sure the PR still adheres to the CONTRIBUTING.md.

SuperSandro2000 commented on pkgs/tools/video/xjadeo/default.nix:
```diff
@@ -0,0 +1,47 @@
+{ lib, stdenv, autoreconfHook, fetchFromGitHub, ffmpeg, freetype, libGLU
+, libjack2, liblo, libX11, libXv, pkg-config, portmidi, xorg }:
+stdenv.mkDerivation rec {
```
```suggestion
, libjack2, liblo, libX11, libXv, pkg-config, portmidi, xorg }:

stdenv.mkDerivation rec {
```

mitchmindtree replied:
It's a little unclear to me why there should be two newlines between a function's arguments and its body? Is this your opinion, or is this a formatting standard that's documented somewhere? `nixfmt` for example does not seem to mind about the lack of gap.

SuperSandro2000 replied:
If you take a look at other new packages you will see that this is standard formatting. Look at any good example at https://nixos.org/manual/nixpkgs/unstable/#how-to-develop and you will see that it is standard.

P.S.: It does not really matter what nixfmt says and if it minds the gap or not.

SuperSandro2000 commented on pkgs/tools/video/xjadeo/default.nix:
```diff
@@ -0,0 +1,47 @@
+    owner = "x42";
+    repo = "xjadeo";
+    rev = "v${version}";
+    sha256 = "0cwh3zimrysyank3x1bcysdz59f8srqdca7jwv97ri3y63dlhifj";
```
```suggestion
    sha256 = "0dma4cjgbrpy16x63zvfr0xss4lryl0zw7nvixvhq2f6z8day1ds";
```

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 115121` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>

  <summary>1 package failed to build and already failed to build on hydra master:</summary>
  <ul>
<li>xjadeo: log was empty</li>
  </ul>
</details>
<details>
<summary>
2 packages built:
</summary>
<ul>
<li>ardour</li>
<li>harvid</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
ardour:
<details>

An avoidable string conversion got detected: `    printf '#include "libs/ardour/ardour/revision.h"\nnamespace ARDOUR { const char* revision = "${version}"; const char* date = ""; }\n' > libs/ardour/revision.cc`
Please do not convert variables to a string without modifying them but use them directly instead.


warning: attribute-typo
A likely typo in `wafConfigureFlags` argument was found, did you mean `configureFlags`?

Near pkgs/applications/audio/ardour/default.nix:131:3:

```
    |
131 |   wafConfigureFlags = [
    |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/attribute-typo.md
warning: unclear-gpl
`gpl2` is a deprecated license, check if project uses `gpl2Plus` or `gpl2Only` and change `meta.license` accordingly.

Near pkgs/applications/audio/ardour/default.nix:186:5:

```
    |
186 |     license = licenses.gpl2;
    |     ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md
warning: build-tools-in-build-inputs
makeWrapper is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/applications/audio/ardour/default.nix:83:3:

```
   |
83 |   buildInputs = [
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md
</details>
</details>

mitchmindtree commented:
OK, second round of review should now be addressed.

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 115121` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
3 packages built:
</summary>
<ul>
<li>ardour</li>
<li>harvid</li>
<li>xjadeo</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
ardour:
<details>

An avoidable string conversion got detected: `    printf '#include "libs/ardour/ardour/revision.h"\nnamespace ARDOUR { const char* revision = "${version}"; const char* date = ""; }\n' > libs/ardour/revision.cc`
Please do not convert variables to a string without modifying them but use them directly instead.


warning: attribute-typo
A likely typo in `wafConfigureFlags` argument was found, did you mean `configureFlags`?

Near pkgs/applications/audio/ardour/default.nix:131:3:

```
    |
131 |   wafConfigureFlags = [
    |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/attribute-typo.md
warning: unclear-gpl
`gpl2` is a deprecated license, check if project uses `gpl2Plus` or `gpl2Only` and change `meta.license` accordingly.

Near pkgs/applications/audio/ardour/default.nix:186:5:

```
    |
186 |     license = licenses.gpl2;
    |     ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md
</details>
</details>

mitchmindtree commented on pkgs/tools/video/harvid/default.nix:
```diff
@@ -0,0 +1,41 @@
+
+  buildInputs = [ ffmpeg libjpeg libpng ];
+
+  makeFlags = [ "DESTDIR=$(out)" ''libdir="/lib"'' ];
```
This is another case where I wrote it this way originally, but `nixfmt` changed it *to* this.

You mention that it doesn't really matter what nixfmt thinks, but from my understanding the whole point of using a tool with some sort of standard formatting like that is so that we don't have to waste time bike-shedding over styling like we have here and [here](https://github.com/NixOS/nixpkgs/pull/115121#discussion_r589720640).

It seems that I was mistaken in assuming that `nixfmt` (a package in this repository) was some sort of standard tool in the same way that rustfmt is to rust. That said, you seem to have some auto-formatting check or a specific styling in mind. Can you point me to a tool you're using or a document I can refer to so that we can avoid the back and forth on formatting nits?

SuperSandro2000 replied:
> This is another case where I wrote it this way originally, but nixfmt changed it to this.

Yet another reason to use nixpkgs-fmt.

> Can you point me to a tool you're using or a document I can refer to so that we can avoid the back and forth on formatting nits?

nixpkgs-fmt does not enforce such formatting changes but it will also not revert the current way.

Double single quotes is reserved for multiline strings and single line strings should use double quotes.
```
''
...
''

"..."
```

mitchmindtree replied:
> Yet another reason to use nixpkgs-fmt.

After a week, this is the first comment where you've mentioned `nixpkgs-fmt` in review of this PR at all. I did not realise the tool even existed until yesterday.

I personally do not care which tool is used at all - just as long as there is some standard one so that I do not have to guess which maintainer I'm going to get and decipher their personal styling preferences by trawling through the recent PRs they've landed.

> nixpkgs-fmt does not enforce such formatting changes but it will also not revert the current way.

I highly recommend that if you have no direct reference in the standard documentation or standard automated tool that users have access to, that you be more lenient in PRs on these finer styling nits. Otherwise, if it means enough to you, I recommend updating the necessary tools, contributing docs and/or manual accordingly.

mitchmindtree committed:
ardour: clarify gpl2 license with gpl2Plus

mitchmindtree commented:
OK, I've addressed all mentioned formatting nits and added a new commit changing the ardour meta license attribute from gpl2 to gpl2Plus as requested by the automated check.

risicle commented:
Builds for me, both with video enabled and disabled, linux x86_64.

SuperSandro2000 commented on pkgs/applications/audio/ardour/default.nix:
```diff
@@ -158,6 +163,10 @@ stdenv.mkDerivation rec {
     done
     install -vDm 644 "ardour.1"* -t "$out/share/man/man1"
+  '' + lib.optionalString videoSupport ''
+    # `harvid` and `xjadeo` must be accessible in `PATH` for video to work.
```
```suggestion
    # `harvid` and `xjadeo` must be accessible in $PATH for video to work.
```
